### PR TITLE
Dev guide v2 schema design review 1

### PR DESCRIPTION
**Pull request by ericharmeling**
Opening this PR to get a review on the first few pages of the "Design a Database Schema" chapter of the new Developer Guide.

Please review the following files:
- `v21.1/schema-design-overview.md`
- `v21.1/schema-design-database.md`
- `v21.1/schema-design-schema.md`

This PR also includes minor ToC update and a rebase of `master`, so you can ignore all other file changes.

I'm hoping to open another PR tomorrow, with a page on tables and a page on indexes.

### CODE REVIEW

File: v21.1/schema-design-database.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+(4 rows)
+~~~
+
+To follow [authorization best practices](authorization.html#authorization-best-practices), after you create a new database, you should also create a new SQL user to manage the lower-level objects in the new database.
+
+In the open SQL shell, execute the following `CREATE USER` statement:
+
+{% include copy-clipboard.html %}
+~~~ sql
+> CREATE USER max;
```
Comment (ianjevans) [inline v21.1/schema-design-database.md:105]: Minor nit here, but in some of the Hello World examples, the username is `maxroach`.

File: v21.1/schema-design-overview.md [hunk 2]
```diff
@@ -1,7 +1,112 @@
...
+
+### Databases
+
+Database objects make up the first level of the [CockroachDB naming hierarchy](sql-name-resolution.html#naming-hierarchy). They contain [schemas](#schemas).
+
+All CockroachDB clusters include a preloaded database named `defaultdb`. Rather than using the `defaultdb` database, we recommend creating your own database.
+
+For guidance on creating databases, see [Create a Database](schema-design-database.html).
+
+{{site.data.alerts.callout_info}}
```
Comment (ianjevans) [inline v21.1/schema-design-overview.md:25]: This callout is a good candidate for either an include or a fragment include.

File: v21.1/schema-design-overview.md [hunk 3]
```diff
@@ -1,7 +1,112 @@
...
+
+Tables contain *rows* of data. Each value in a row of data belongs to a particular *column*. Each column allows values of data of a single data type. Columns can be further qualified with [column-level constraints](column-constraints.html), or computed with [scalar expressions](computed-columns.html).
+
+For guidance on defining tables, see [Tables](schema-design-tables.html).
+
+### Indexes
+
+Indexes, like tables, belong to a single user-defined schema, in the third level of the naming hierarchy.
+
+An index is a copy of the rows in a single table, sorted by a column or set of columns. CockroachDB queries use indexes to more efficiently find data in a table, given the rows values of a particular column.
```
Comment (ianjevans) [inline v21.1/schema-design-overview.md:53]:
```suggestion
An index is a copy of the rows in a single table, sorted by a column or set of columns. CockroachDB queries use indexes to more efficiently find data in a table, given the values of a particular column.
```

File: v21.1/schema-design-overview.md [hunk 4]
```diff
@@ -1,7 +1,112 @@
...
+
+The two main types of indexes are the `primary` index, an index on the row-identifying [primary key columns](primary-key.html), and the secondary index, an index created on columns of your choice.
```
Comment (ianjevans) [inline v21.1/schema-design-overview.md:55]:
```suggestion
The two main types of indexes are the `primary` index, an index on the row-identifying [primary key columns](primary-key.html), and the secondary index, an index created on any other non-primary key column.
```

File: v21.1/schema-design-overview.md [hunk 5]
```diff
@@ -1,7 +1,112 @@
...
+
+For guidance on defining primary keys, see [Select Primary Key Columns](schema-design-tables.html#select-primary-key-columns). For guidance on defining secondary indexes, see [Add a Secondary Index](schema-design-indexes.html).
+
+#### Specialized indexes
+
+CockroachDB supports some specialized types of indexes, designed to improve query performance in specific use cases. For guidance on specialized indexes, see [Index a Subset of Rows](partial-indexes.html) and [Index Sequential Keys](hash-sharded-indexes.html).
```
Comment (ianjevans) [inline v21.1/schema-design-overview.md:61]: I'm not totally sure where inverted indexes figure into this discussion, but just want to note them.

File: v21.1/schema-design-overview.md [hunk 6]
```diff
@@ -1,7 +1,112 @@
...
 toc: true
 ---
 
-WIP
\ No newline at end of file
+This page provides an overview of database schemas in CockroachDB.
+
+In later [**Design a Database Schema** sections](schema-design-database.html), we provide best practices for designing a database schema that optimizes performance and storage resources.
+
+## Database Schema Objects
```
Comment (ianjevans) [inline v21.1/schema-design-overview.md:11]: Is our in-house style sentence case or capital case for headings?

File: v21.1/schema-design-schema.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+- When you create a user-defined schema, take note of the [object's *owner*](authorization.html#object-ownership). You can specify the owner in a `CREATE SCHEMA` statement with the `AUTHORIZATION` keyword. If `AUTHORIZATION` is not specified, the owner will be the user creating the user-defined schema. We recommend creating user-defined schemas as the user that will create and use the lower-level objects of the database schema (e.g., tables and indexes).
+
+{% include {{page.version.version}}/sql/dev-schema-changes.md %}
+
+### Example
+
+{{site.data.alerts.callout_info}}
+For the examples in the rest of the [**Design a Database Schema** sections](schema-design-table.html), we'll add all database schema change statements to a `.sql` file, and then pass the file to the [SQL client](cockroach-sql.html) for execution as the `max` user.
```
Comment (ianjevans) [inline v21.1/schema-design-schema.md:58]: See previous comment on the `max` vs. `maxroach` user.

**Review by ericharmeling — COMMENTED** [review]
TFTR, @ianjevans ! 

I think I got to all of your comments. 

I also forgot to add an include (`_includes/v21.1/sql/dev-schema-changes.md`) earlier.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/9487#-:-MRzBglc8OiZxdiDs7RR:bunl0dr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ianjevans)

___
*[v21.1/schema-design-database.md, line 105 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehme-r1-105:-MRvL-0c4-jDmcv1_Q0k:buw2ipt) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-database.md#L105)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

Minor nit here, but in some of the Hello World examples, the username is `maxroach`.
</blockquote></details>

You're right. I like `maxroach` better. Changed.

FWIW, we do use `max` in some examples as well (mainly SQL reference examples).
___
*[v21.1/schema-design-overview.md, line 11 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmf-r1-11:-MRvLMaJ-izdQi29lDOt:b-v7znyt) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-overview.md#L11)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

Is our in-house style sentence case or capital case for headings?
</blockquote></details>

Sentence case. Good catch! Fixed.
___
*[v21.1/schema-design-overview.md, line 25 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmf-r1-25:-MRyxqvD62SJle4vzTnZ:b7nkff1) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-overview.md#L25)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

This callout is a good candidate for either an include or a fragment include.
</blockquote></details>

Good idea. I made it into an include and added it to the databases page.

Did the same for schemas too.
___
*[v21.1/schema-design-overview.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmf-r1-53:-MRyz94R6TvxU9qfQXh-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-overview.md#L53)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

```suggestion
An index is a copy of the rows in a single table, sorted by a column or set of columns. CockroachDB queries use indexes to more efficiently find data in a table, given the values of a particular column.
```
</blockquote></details>

Done.
___
*[v21.1/schema-design-overview.md, line 55 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmf-r1-55:-MRyzKpC9Peciebqyew3:b80polw) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-overview.md#L55)):*
> any other non-primary key column.

I'm pretty sure you are allowed to index primary key columns in a secondary index. If both are composite indexes, it might make sense to have a secondary index with a subset or superset of the columns in the primary key.
___
*[v21.1/schema-design-overview.md, line 61 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmf-r1-61:-MRz8bgS-bebORfe7k1e:b6sy3ns) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-overview.md#L61)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

I'm not totally sure where inverted indexes figure into this discussion, but just want to note them.
</blockquote></details>

Good point. I updated this section (and the ToC) to reference the pages on inverted indexes and spatial indexes.

Just calling those headers "Inverted Indexes" and "Spatial Indexes" seems a little too esoteric for developers, so I'm calling them "Index JSON and Array Data" (inverted-indexes.html) and "Index Spatial Data" (spatial-indexes.html).
___
*[v21.1/schema-design-schema.md, line 58 at r1](https://reviewable.io/reviews/cockroachdb/docs/9487#-MRuRib8jWA3gLt6Ehmg-r1-58:-MRzBK0w598g_d4yNFTg:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/3f35e8eb0b77e2339329543a39143e337b24d7c9/v21.1/schema-design-schema.md#L58)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

See previous comment on the `max` vs. `maxroach` user.
</blockquote></details>

Done.

**Review by ianjevans — APPROVED** [review]

File: v21.1/schema-design-schema.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+User-defined schemas belong to the second level of the [CockroachDB naming hierarchy](sql-name-resolution.html#naming-hierarchy).
+
+To create a user-defined schema, use a [`CREATE SCHEMA` statement](create-schema.html), following [the user-defined schema best practices](#user-defined-schema-best-practices). After reviewing the best practices, see the example we provide [below](#example).
+
+### User-defined schema best practices
+
+Here are some best practices to follow when creating and using user-defined schemas:
+
+- Do not use the preloaded `public` schema. Instead, create a user-defined schema with `CREATE SCHEMA`, and then refer to lower-level objects with fully-qualified names (e.g., `schema_name.table_name`).
```
Comment (ericharmeling) [inline v21.1/schema-design-schema.md:48]: @vy-ton 

Following up on the SQL product area meeting, this is the guidance we currently give re: user-defined schemas. We can remove this recommendation, or we can reword it to recommend that users create UDS if they fit their use case (e.g., different users).

Comment (vy-ton) [reply v21.1/schema-design-schema.md:48]: I think we should reword. There's no reason to avoid the `public` schema. Database objects can also be moved between schemas as long as there's no name collision. So you can start with `public` and then move to user-defined schemas as needed.

What we want to emphasize is that UDS provides another level that can be secured. And to showcase that with an example like users. Sometimes tests would execute in their own UDS as well

**Comment by ericharmeling** [comment]
> I'm pretty sure you are allowed to index primary key columns in a secondary index. If both are composite indexes, it might make sense to have a secondary index with a subset or superset of the columns in the primary key.

I was wrong about this. Primary key columns are implicit in all secondary index keys... a fact that we do not document very well! And we actually have a few examples (multi-region, in particular) that create secondary indexes with region-indicating columns that are in the primary key.
